src/runner.jl

```diff
@@ -167,28 +175,66 @@ function test!(pkg::AbstractString,
-        runner_file_path = make_runner_file(testfilepath, logfilename)
-        cd(dirname(testfilepath)) do
-            try
-                cmd = ```
-                    $(Base.julia_cmd())
-                    --code-coverage=$(coverage ? "user" : "none")
-                    --color=$(Base.have_color ? "yes" : "no")
-                    --compiled-modules=$(Bool(Base.JLOptions().use_compiled_modules) ? "yes" : "no")
-                    --check-bounds=yes
-                    --depwarn=$(Base.JLOptions().depwarn == 2 ? "error" : "yes")
-                    --inline=$(Bool(Base.JLOptions().can_inline) ? "yes" : "no")
-                    --startup-file=$(Base.JLOptions().startupfile != 2 ? "yes" : "no")
-                    --track-allocation=$(("none", "user", "all")[Base.JLOptions().malloc_log + 1])
-                    $(runner_file_path)
-                    ```
-                run(cmd)
-                @info "$pkg tests passed. Results saved to $logfilename."
-            catch err
-                @warn "ERROR: Test(s) failed or had an error in $pkg"
-                push!(errs,pkg)
+        testreportsdir = dirname(@__DIR__)
+        runner_code = gen_runner_code(testfilepath, logfilename, testreportsdir)
+        cmd = ```
+            $(Base.julia_cmd())
+            --code-coverage=$(coverage ? "user" : "none")
+            --color=$(Base.have_color ? "yes" : "no")
+            --compiled-modules=$(Bool(Base.JLOptions().use_compiled_modules) ? "yes" : "no")
+            --check-bounds=yes
+            --depwarn=$(Base.JLOptions().depwarn == 2 ? "error" : "yes")
+            --inline=$(Bool(Base.JLOptions().can_inline) ? "yes" : "no")
+            --startup-file=$(Base.JLOptions().startupfile != 2 ? "yes" : "no")
+            --track-allocation=$(("none", "user", "all")[Base.JLOptions().malloc_log + 1])
+            --eval $(runner_code)
+            ```
+
+        test_folder_has_project_file = isfile(replace(testfilepath, "runtests.jl"=>"Project.toml"))
+
+        if VERSION >= v"1.4.0" || (VERSION >= v"1.2.0" && test_folder_has_project_file)
+            # Operations.sandbox() has different arguments between versions
+            @static if VERSION >= v"1.4.0"
+                test_project_override = test_folder_has_project_file ?
+                    nothing :
+                    gen_target_project(ctx, pkgspec, pkgspec.path, "test")
+                sandbox_args = (ctx,
+                                pkgspec,
+                                pkgspec.path,
+                                joinpath(pkgspec.path, "test"),
+                                test_project_override)
+            else
+                sandbox_args = (ctx,
+                                pkgspec,
+                                pkgspec.path,
+                                joinpath(pkgspec.path, "test"))
+            end
+
+            sandbox(sandbox_args...) do
```

Comment:
shorter and while technically slower, negligably so.
```suggestion
            sandbox_args = (ctx,
                            pkgspec,
                            pkgspec.path,
                            joinpath(pkgspec.path, "test"))   
            if VERSION >= v"1.4.0" && !test_folder_has_project_file
                test_project_override = gen_target_project(ctx, pkgspec, pkgspec.path, "test")
                sandbox_args = (sandbox_args..., test_project_override)
            end

            sandbox(sandbox_args...) do
```

Reply:
Yeah that's cleaner. The inner `if` condition still needs to be

```Julia
if VERSION >= v"1.4.0" 
    test_project_override = test_folder_has_project_file ?
        nothing :
        gen_target_project(ctx, pkgspec, pkgspec.path, "test")
    sandbox_args = (sandbox_args..., test_project_override)
end
```
as `sandbox` needs the fifth argument in V1.4, so I've merged your suggestion and that.